maven/impl-maven-embedded/src/main/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/DistributionStageImpl.java

```diff
@@ -49,29 +45,30 @@ public NEXT_STEP useDistribution(URL mavenDistribution, boolean useCache) {
 
         File mavenDir = prepareMavenDir(useCache);
         File downloaded = download(mavenDir, mavenDistribution);
-
         String downloadedZipMd5hash = getMd5hash(downloaded);
         File withExtractedDir = null;
-        boolean isNewlyExtracted = false;
         if (downloadedZipMd5hash != null) {
             withExtractedDir = checkIfItIsAlreadyExtracted(downloadedZipMd5hash);
-        }
-        if (withExtractedDir == null) {
-            withExtractedDir = extract(downloaded);
-            isNewlyExtracted = true;
-        }
-        File binDirectory = retrieveBinDirectory(withExtractedDir);
+            if (withExtractedDir == null) {
+                withExtractedDir = extract(downloaded, downloadedZipMd5hash);
+            }
+            File binDirectory = retrieveBinDirectory(withExtractedDir);
 
-        if (isNewlyExtracted) {
-            addMd5hashFile(downloadedZipMd5hash, withExtractedDir);
+            useInstallation(binDirectory);
         }
 
-        useInstallation(binDirectory);
         return returnNextStepType();
     }
 
     private File checkIfItIsAlreadyExtracted(final String downloadedZipMd5hash) {
-        File[] dirs = new File(mavenTargetDir).listFiles(new FilterDirWithMd5Hash(downloadedZipMd5hash));
+
+        File[] dirs = new File(mavenTargetDir).listFiles(new FileFilter() {
+
+            @Override
+            public boolean accept(File file) {
+                    return file.isFile() && downloadedZipMd5hash.equals(file.getName());
```

Comment:
`file.isFile()` I think that here you should check that it is a directory, not a file. In this case, nothing will be matched and the zip file will be extracted again and the current dir will be replaced by the new one...